Geometry/HGCalCommonData/src/HGCalWaferType.cc

```diff
@@ -30,28 +33,61 @@ HGCalWaferType::HGCalWaferType(const std::vector<double>& rad100,
 HGCalWaferType::~HGCalWaferType() { }
 
 int HGCalWaferType::getType(double xpos, double ypos, double zpos) {
-  double xc[HGCalParameters::k_CornerSize], yc[HGCalParameters::k_CornerSize];
+  std::vector<double> xc(HGCalParameters::k_CornerSize,0);
+  std::vector<double> yc(HGCalParameters::k_CornerSize,0);
   xc[0] = xpos+r_;  yc[0] = ypos+0.5*R_;
   xc[1] = xpos;     yc[1] = ypos+R_;
   xc[2] = xpos-r_;  yc[2] = ypos+0.5*R_;
   xc[3] = xpos-r_;  yc[3] = ypos-0.5*R_;
   xc[4] = xpos;     yc[4] = ypos-R_;
   xc[5] = xpos+r_;  yc[5] = ypos-0.5*R_;
   std::pair<double,double> rv = rLimits(zpos);
-  int fine(0), coarse(0);
+  std::vector<int>         fine, coarse;
   for (unsigned int k=0; k<HGCalParameters::k_CornerSize; ++k) {
     double rpos = std::sqrt(xc[k]*xc[k]+yc[k]*yc[k]);
-    if      (rpos <= rv.first)  ++fine;
-    else if (rpos <= rv.second) ++coarse;
+    if      (rpos <= rv.first)  fine.push_back(k);
+    else if (rpos <= rv.second) coarse.push_back(k);
+  }
+  int    type(-2);
+  double fracArea(0);
+  if (choice_ == 1) {
+    if      (fine.size()   >= cutValue_) type = 0;
+    else if (coarse.size() >= cutValue_) type = 1;
+    else                                 type = 2;
+  } else {
+    if      (fine.size()   >= 4)                     type = 0;
+    else if (coarse.size() >= 4 && fine.size() <= 1) type = 1;
+    else if (coarse.size() <  2 && fine.empty())     type = 2;
+    else if (!fine.empty())                          type =-1;
+    if (type < 0) {
+      unsigned int kmax = (type == -1) ? fine.size() : coarse.size();
+      std::vector<double> xcn, ycn;
+      for (unsigned int k=0; k<kmax; ++k) {
+	unsigned int k1 = (type == -1) ? fine[k] : coarse[k];
+	unsigned int k2 = (k1 == xc.size()-1) ? 0 : k1+1;
+	bool ok = ((type == -1) ?
+		   (std::find(fine.begin(),fine.end(),k2) != fine.end()) :
+		   (std::find(coarse.begin(),coarse.end(),k2) != coarse.end()));
+	if (!ok) {
+	  xcn.emplace_back(xc[k1]); ycn.emplace_back(yc[k1]);
```

Comment:
it seems like this line is done in either case (`ok` or not). Is that the correct behavior? If so, it could be moved outside of the conditional to reduce code duplication.